Source/CesiumRuntime/Private/CesiumEllipsoid.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020-2024 CesiumGS, Inc. and Contributors
+
+#include "CesiumEllipsoid.h"
+#include "CesiumEllipsoidFunctions.h"
+#include "CesiumRuntime.h"
+#include "VecMath.h"
+
+#include "MathUtil.h"
+#include "UObject/ConstructorHelpers.h"
+
+#include <CesiumGeospatial/Ellipsoid.h>
+
+using namespace CesiumGeospatial;
+
+FVector UCesiumEllipsoid::GetRadii() { return this->Radii; }
+
+void UCesiumEllipsoid::SetRadii(FVector NewRadii) { this->Radii = NewRadii; }
+
+double UCesiumEllipsoid::GetMaximumRadius() { return this->Radii.X; }
+
+double UCesiumEllipsoid::GetMinimumRadius() { return this->Radii.Z; }
+
+FVector UCesiumEllipsoid::ScaleToGeodeticSurface(
+    const FVector& EllipsoidCenteredEllipsoidFixedPosition) {
+  return CesiumEllipsoidFunctions::ScaleToGeodeticSurface(
+      *this->GetNativeEllipsoid(),
+      EllipsoidCenteredEllipsoidFixedPosition);
+}
+
+FVector UCesiumEllipsoid::GeodeticSurfaceNormal(
+    const FVector& EllipsoidCenteredEllipsoidFixedPosition) {
+  return CesiumEllipsoidFunctions::GeodeticSurfaceNormal(
+      *this->GetNativeEllipsoid(),
+      EllipsoidCenteredEllipsoidFixedPosition);
+}
+
+FVector
+UCesiumEllipsoid::LongitudeLatitudeHeightToEllipsoidCenteredEllipsoidFixed(
+    const FVector& LongitudeLatitudeHeight) {
+  return CesiumEllipsoidFunctions::
+      LongitudeLatitudeHeightToEllipsoidCenteredEllipsoidFixed(
+          *this->GetNativeEllipsoid(),
+          LongitudeLatitudeHeight);
+}
+
+FVector
+UCesiumEllipsoid::EllipsoidCenteredEllipsoidFixedToLongitudeLatitudeHeight(
+    const FVector& EllipsoidCenteredEllipsoidFixedPosition) {
+  return CesiumEllipsoidFunctions::
+      EllipsoidCenteredEllipsoidFixedToLongitudeLatitudeHeight(
+          *this->GetNativeEllipsoid(),
+          EllipsoidCenteredEllipsoidFixedPosition);
+}
+
+FMatrix UCesiumEllipsoid::EastNorthUpToEllipsoidCenteredEllipsoidFixed(
+    const FVector& EllipsoidCenteredEllipsoidFixedPosition) {
+  return CesiumEllipsoidFunctions::EastNorthUpToEllipsoidCenteredEllipsoidFixed(
+      *this->GetNativeEllipsoid(),
+      EllipsoidCenteredEllipsoidFixedPosition);
+}
+
+LocalHorizontalCoordinateSystem
+UCesiumEllipsoid::CreateCoordinateSystem(const FVector& Center, double Scale) {
+  {
+    return LocalHorizontalCoordinateSystem(
+        VecMath::createVector3D(Center),
+        LocalDirection::East,
+        LocalDirection::South,
+        LocalDirection::Up,
+        1.0 / Scale,
+        *this->GetNativeEllipsoid());
+  }
+}
+
+TSharedPtr<Ellipsoid> UCesiumEllipsoid::GetNativeEllipsoid() {
+  const double MinRadiiValue = TMathUtilConstants<double>::Epsilon;
+
+  if (!this->NativeEllipsoid.IsValid()) {
+    // Radii of zero will throw Infs and NaNs into our calculations which will
+    // cause Unreal to crash when the values reach a transform.
+    if (this->Radii.X < MinRadiiValue || this->Radii.Y < MinRadiiValue ||
```

Comment:
Looks like a good check. 

Optionally could be moved to `::SetRadii` where we could log the bad input, like we do here, and even reject making the change to the member variables.

Reply:
The reason it's here is because, when you create a new Ellipsoid asset, the Radii field is initialized to (0, 0, 0) by default. If you pass this asset to a UCesiumGeoreference without changing the values, it'll crash Unreal (and if you saved the scene, it'll crash Unreal every time you open the scene). Could initialize `UCesiumEllipsoid::Radii` to something other than (0, 0, 0), but I think it's better to keep the check here.